### ISSUE TITLE
AddControllersAsServices only works if controllers are registered first

### ISSUE THREAD

**Issue by YairHalberstadt** [dotnet/aspnetcore#27959]
### Describe the bug
Using `AddControllersAsServices()` only uses controller registrations to resolve a controller if the controller is registered before `AddControllersAsServices()`, not after. Otherwise it just tries to resolve using the constructor.

### To Reproduce

The following single file program runs without error in .Net 5.0

```csharp
using Microsoft.AspNetCore.Builder;
using Microsoft.AspNetCore.Hosting;
using Microsoft.AspNetCore.Mvc;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Hosting;

Host.CreateDefaultBuilder()
    .ConfigureWebHostDefaults(webBuilder =>
    {
        webBuilder.UseStartup<Startup>();
    }).Build().Run();

public class Startup
{
    public Startup(IConfiguration configuration)
    {
        Configuration = configuration;
    }

    public IConfiguration Configuration { get; }

    public void ConfigureServices(IServiceCollection services)
    {
        services.AddTransient(x => new Controller("Hello World"));
        services.AddControllers().AddControllersAsServices();
    }

    public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
    {
        app.UseRouting();

        app.UseEndpoints(endpoints =>
        {
            endpoints.MapControllers();
        });
    }
}

[ApiController]
[Route("/")]
public class Controller : ControllerBase
{
    private readonly string _str;

    public Controller(string str)
    {
        _str = str;
    }

    [HttpGet]
    public string Get() => _str;
}
```

When we switch around the lines in `ConfigureServices` though:

```csharp
    public void ConfigureServices(IServiceCollection services)
    {
        services.AddControllers().AddControllersAsServices();
        services.AddTransient(x => new Controller("Hello World"));
    }
```

we get the following exception:

```
System.AggregateException: 'Some services are not able to be constructed (Error while validating the service descriptor 'ServiceType: Controller Lifetime: Transient ImplementationType: Controller': Unable to resolve service for type 'System.String' while attempting to activate 'Controller'.)'
```

Note that when explicitly calling `services.BuildServiceProvider().GetRequiredService<Controller>();` we successfully construct `Controller` either way.

### Further technical details
- ASP.NET Core version 5.0
- VS 16.9 preview 1

**Comment by javiercn** [comment]
@YairHalberstadt thanks for contacting us.

That's expected and by design, the issue you are facing is that you are overriding the existing registration that `AddControllersAsServices` makes. So you should be using `TryAddTransient` if you don't want to override the registration within the DI container.

**Comment by YairHalberstadt** [comment]
@javiercn 

It's the other way round - I dont want AddControllersAsServices to overwrite my registrations. I need to use AddControllersAsServices because otherwise it resolves controllers without using the Service provider at all.

**Comment by YairHalberstadt** [comment]
Note that when we call `services.BuildServiceProvider().GetRequiredService<Controller>();` it always uses the service descriptor I registered explicitly, so I'm not sure why it behaves differently when being created by Asp.Net.

**Comment by javiercn** [comment]
The default controller activator has a different behavior than the service based controller activator.

**Comment by YairHalberstadt** [comment]
@javiercn 

I found the real problem.

When I disable "ValidateOnBuild":

```csharp
Host.CreateDefaultBuilder()
    .ConfigureWebHostDefaults(webBuilder =>
    {
        webBuilder.UseStartup<Startup>();
    })
    .UseDefaultServiceProvider((context, options) =>
    {
        options.ValidateOnBuild = false;
    })
    .Build()
    .Run();
```

Everything runs fine, and the registration I provided explicitly is used.
However if when `ValidateOnBuild` is true, it validates all registrations, even those which will not in fact be used, and so crashes because `AddControllersAsServices` adds a registration for `Controller` which cannot be resolved.

Whilst unfortunate, I understand that this is by design and cannot be changed.
